provider/src/test/kotlin/org/gradle/kotlin/dsl/provider/ScriptApiTest.kt

```diff
@@ -0,0 +1,165 @@
+    when (this) {
+        is KFunction -> isCompatibleWith(api)
+        is KProperty -> isCompatibleWith(api)
+        else         -> false
+    }
+
+
+private
+fun KProperty<*>.isCompatibleWith(api: KCallable<*>) =
+    this::class == api::class && name == api.name && returnType == api.returnType
+
+
+private
+fun KFunction<*>.isCompatibleWith(api: KCallable<*>) =
+    when {
+        api is KProperty && api !is KMutableProperty && isCompatibleWithGetterOf(api) -> true
+        api is KFunction && isCompatibleWith(api)                                     -> true
+        else                                                                          -> false
+    }
+
+
+private
+fun KFunction<*>.isCompatibleWithGetterOf(api: KProperty<*>) =
+    name == api.javaGetter?.name
+        && returnType == api.getter.returnType
+        && valueParameters.isEmpty() && api.getter.valueParameters.isEmpty()
+
+
+private
+fun KFunction<*>.isCompatibleWith(api: KFunction<*>) =
+    name == api.name && returnType == api.returnType && valueParameters.isCompatibleWith(api.valueParameters)
+
+
+private
+fun List<KParameter>.isCompatibleWith(api: List<KParameter>) =
+    when {
+        size != api.size -> false
+        isEmpty()        -> true
+        else             -> (0..(size - 1)).all { idx -> this[idx].isCompatibleWith(api[idx]) }
+    }
+
+
+private
+fun KParameter.isCompatibleWith(api: KParameter) =
+    when {
+        isVarargCompatibleWith(api)       -> true
+        isGradleActionCompatibleWith(api) -> true
+        type == api.type                  -> true
+        else                              -> false
+    }
+
+
+private
+fun KParameter.isGradleActionCompatibleWith(api: KParameter) =
+    type.jvmErasure == Action::class
+        && api.isSamWithReceiverReturningUnit()
+        && type.arguments[0].type!!.isTypeArgumentCompatibleWith(api.type.arguments[0].type!!)
+
+
+private
+fun KParameter.isSamWithReceiverReturningUnit() =
+    type.jvmErasure == Function1::class
+        && type.arguments[1] == KTypeProjection(KVariance.INVARIANT, Unit::class.createType())
+
+
+private
+fun KParameter.isVarargCompatibleWith(api: KParameter) =
+    isVararg && api.isVararg && type.isParameterTypeCompatibleWith(api.type)
+
+
+private
+fun KType.isParameterTypeCompatibleWith(apiParameterType: KType) =
+    when {
+        this == apiParameterType                     -> true
+        classifier != apiParameterType.classifier    -> false
+        hasCompatibleTypeArguments(apiParameterType) -> true
+        else                                         -> false
+    }
+
+
+private
+fun KType.hasCompatibleTypeArguments(api: KType) =
+    arguments.size == api.arguments.size && (0..(arguments.size - 1)).all { idx ->
+        arguments[idx].type!!.isTypeArgumentCompatibleWith(api.arguments[idx].type!!)
+    }
+
+
+private
+fun KType.isTypeArgumentCompatibleWith(api: KType) =
+    withNullability(false) == api
```

Comment:
This is very naive but works for currently covered use cases.
I couldn't find a way to distinguish `Any?` from `Any!` nor investigated querying jsr-305.